wikidata/search.py

```diff
@@ -0,0 +1,415 @@
+    :param entity_type: Optional filter for entity type. Can be 'person', 'place', 'organization',
+                        'work', etc. Filters results based on description content
+    :type entity_type: :class:`str`
+    :param likeness_threshold: Minimum similarity score threshold (0-1) for results to include.
+                              For example, 0.6 means only return results with 60% or higher similarity
+                              (default: 1.0, returns only exact matches)
+    :type likeness_threshold: :class:`float`
+    :param verbose: If True, pretty print the search results to stdout (default: False)
+    :type verbose: :class:`bool`
+    :return: A list of SearchResult objects sorted by similarity score (highest first),
+             filtered by likeness_threshold if provided
+    :rtype: :class:`List`\\ [:class:`SearchResult`]
+    :raises urllib.error.URLError: If the HTTP request fails
+    
+    """
+    # Validate input
+    if not query or not isinstance(query, str):
+        return []
+
+    if limit < 1:
+        limit = 1
+    
+    # Validate likeness_threshold
+    if not 0.0 <= likeness_threshold <= 1.0:
+        likeness_threshold = 1.0
+
+    # Determine which base_url and opener to use
+    if client is not None:
+        search_base_url = client.base_url
+        http_opener = client.opener
+    else:
+        search_base_url = WIKIDATA_BASE_URL
+        http_opener = opener
+
+    # Build search URL
+    search_url = f'{search_base_url}w/index.php?search={urllib.parse.quote(query)}'
+    
+    # Add language parameter if specified
+    if language:
+        search_url += f'&language={urllib.parse.quote(language)}'
+
+    try:
+        # Fetch the search results page
+        req = urllib.request.Request(search_url)
+
+        # Add User-Agent header to avoid 403 Forbidden responses
+        user_agent = (
+            'WikidataClientPython '
+            '(https://github.com/dahlia/wikidata; hong@minhee.org)'
+        )
+        req.add_header('User-Agent', user_agent)
+
+        if http_opener is None:
+            response = urllib.request.urlopen(req)
+        else:
+            response = http_opener.open(req)
+
+        html_content = response.read().decode('utf-8')
+
+    except urllib.error.URLError as e:
+        raise
+
+    # Parse HTML and extract results
+    parser = _SearchResultParser()
+    parser.feed(html_content)
+
+    # Calculate similarity scores
+    for result in parser.results:
+        result.similarity_score = _calculate_similarity(query, result.name)
+
+    # Sort by similarity score (highest first)
+    sorted_results = sorted(
+        parser.results,
+        key=lambda r: r.similarity_score,
+        reverse=True
+    )
+
+    # Filter by entity type if specified
+    if entity_type:
+        sorted_results = [
+            result for result in sorted_results
+            if result.description and entity_type.lower() in result.description.lower()
+        ]
```

Comment:
Filtering by substring match on description is unreliable. For example, searching for `entity_type="person"` would match “personal computer” or “in person.”

The official API has a `type` parameter (`item`, `property`) for this purpose.